In [1]:
import pandas as pd

data=pd.DataFrame.from_csv("winemag-data_first150k.csv")

/Users/aggrace/anaconda2/envs/py37/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: from_csv is deprecated. Please use read_csv(...) instead. Note that some of the default arguments are different, so please refer to the documentation for from_csv when changing your function calls
  This is separate from the ipykernel package so we can avoid doing imports until


In [12]:
data=pd.read_csv('winemag-data_first150k.csv')

In [13]:
data.head()

Unnamed: 0 country                                        description  \
0           0      US  This tremendous 100% varietal wine hails from ...   
1           1   Spain  Ripe aromas of fig, blackberry and cassis are ...   
2           2      US  Mac Watson honors the memory of a wine once ma...   
3           3      US  This spent 20 months in 30% new French oak, an...   
4           4  France  This is the top wine from La Bégude, named aft...   

                            designation  points  price        province  \
0                     Martha's Vineyard      96  235.0      California   
1  Carodorum Selección Especial Reserva      96  110.0  Northern Spain   
2         Special Selected Late Harvest      96   90.0      California   
3                               Reserve      96   65.0          Oregon   
4                            La Brûlade      95   66.0        Provence   

            region_1           region_2             variety  \
0        Napa Valley               Napa  Cabernet Sauvignon   
1               Toro                NaN       Tinta de Toro   
2     Knights Valley             Sonoma     Sauvignon Blanc   
3  Willamette Valley  Willamette Valley          Pinot Noir   
4             Bandol                NaN  Provence red blend   

                    winery  
0                    Heitz  
1  Bodega Carmen Rodríguez  
2                 Macauley  
3                    Ponzi  
4     Domaine de la Bégude

In [15]:
count_row = data.shape[0]
count_row

150930

In [16]:
data.describe()

Unnamed: 0         points          price
count  150930.000000  150930.000000  137235.000000
mean    75464.500000      87.888418      33.131482
std     43569.882402       3.222392      36.322536
min         0.000000      80.000000       4.000000
25%     37732.250000      86.000000      16.000000
50%     75464.500000      88.000000      24.000000
75%    113196.750000      90.000000      40.000000
max    150929.000000     100.000000    2300.000000

In [28]:
df=data[['description','country','designation','variety','price','winery','points']]

In [29]:
df.head()

description country  \
0  This tremendous 100% varietal wine hails from ...      US   
1  Ripe aromas of fig, blackberry and cassis are ...   Spain   
2  Mac Watson honors the memory of a wine once ma...      US   
3  This spent 20 months in 30% new French oak, an...      US   
4  This is the top wine from La Bégude, named aft...  France   

                            designation             variety  price  \
0                     Martha's Vineyard  Cabernet Sauvignon  235.0   
1  Carodorum Selección Especial Reserva       Tinta de Toro  110.0   
2         Special Selected Late Harvest     Sauvignon Blanc   90.0   
3                               Reserve          Pinot Noir   65.0   
4                            La Brûlade  Provence red blend   66.0   

                    winery  points  
0                    Heitz      96  
1  Bodega Carmen Rodríguez      96  
2                 Macauley      96  
3                    Ponzi      96  
4     Domaine de la Bégude      95

In [45]:
df.drop_duplicates(subset=None, keep='first', inplace=False)

description      country  \
0       This tremendous 100% varietal wine hails from ...           US   
1       Ripe aromas of fig, blackberry and cassis are ...        Spain   
2       Mac Watson honors the memory of a wine once ma...           US   
3       This spent 20 months in 30% new French oak, an...           US   
4       This is the top wine from La Bégude, named aft...       France   
5       Deep, dense and pure from the opening bell, th...        Spain   
6       Slightly gritty black-fruit aromas include a s...        Spain   
7       Lush cedary black-fruit aromas are luxe and of...        Spain   
8       This re-named vineyard was formerly bottled as...           US   
9       The producer sources from two blocks of the vi...           US   
10      Elegance, complexity and structure come togeth...        Italy   
11      From 18-year-old vines, this supple well-balan...           US   
12      A standout even in this terrific lineup of 201...           US   
13      This wine is in peak condition. The tannins an...       France   
14      With its sophisticated mix of mineral, acid an...           US   
15      First made in 2006, this succulent luscious Ch...           US   
16      This blockbuster, powerhouse of a wine suggest...           US   
17      Nicely oaked blackberry, licorice, vanilla and...        Spain   
18      Coming from a seven-acre vineyard named after ...       France   
19      This fresh and lively medium-bodied wine is be...           US   
20      Heitz has made this stellar rosé from the rare...           US   
21      Alluring, complex and powerful aromas of grill...        Spain   
22      Tarry blackberry and cheesy oak aromas are app...        Spain   
23      The apogee of this ambitious winery's white wi...           US   
24      San Jose-based producer Adam Comartin heads 1,...           US   
25      Yields were down in 2015, but intensity is up,...  New Zealand   
26      Bergström has made a Shea designate since 2003...           US   
27      Focused and dense, this intense wine captures ...           US   
28      Cranberry, baked rhubarb, anise and crushed sl...           US   
29      This standout Rocks District wine brings earth...           US   
...                                                   ...          ...   
149610  A sadly unbalanced wine from a company that ha...     Portugal   
149611  Lots of clove and oak on the nose, with some g...    Argentina   
149612  Offers perfumed floral/minerally notes on the ...    Australia   
149613  Drinks hot and thick in tannins, and despite s...           US   
149614  Deep gold in color, with nut, peach and overt ...    Australia   
149615  Full plum and raspberry aromas are not overpow...    Argentina   
149616  Not unkind aromas of bath oil and flowers set ...    Argentina   
149617  Bacon, lychee and bubble gum aromas are unexpe...           US   
149618  This blending for the American market is softe...       France   
149619  This is more like an Italian bitters than a Po...     Portugal   
149620  Standard aromas in the berry and beet range is...    Argentina   
149621  Starts off with grassy, fresh herbal aromas an...    Australia   
149622  Yellowish in color, with a heavy, somewhat cre...    Argentina   
149623  Dark and dense, with extracted black cherry, l...    Argentina   
149624  An easy, fresh, ripe style, with dried fruits ...     Portugal   
149625  Tastes like a cool-climate SB with the grapefr...           US   
149626  Made in a superripe style, this Chard is jammy...           US   
149627  Opens with spice aromas that seem a bit unusua...       France   
149628  Big and dark, with smoked meat and bacon aroma...    Argentina   
149629  Half Merlot, with the rest Cabernet Sauvignon,...           US   
149630  Tastes kind of soft and thick in jammy cherry ...           US   
149631  This is a very old 40-year old, with some hars...     Portugal   
149632  Nut and white stone fruit aromas start thi

In [49]:
df1=df.dropna()

In [50]:
df1

description      country  \
0       This tremendous 100% varietal wine hails from ...           US   
1       Ripe aromas of fig, blackberry and cassis are ...        Spain   
2       Mac Watson honors the memory of a wine once ma...           US   
3       This spent 20 months in 30% new French oak, an...           US   
4       This is the top wine from La Bégude, named aft...       France   
5       Deep, dense and pure from the opening bell, th...        Spain   
6       Slightly gritty black-fruit aromas include a s...        Spain   
7       Lush cedary black-fruit aromas are luxe and of...        Spain   
8       This re-named vineyard was formerly bottled as...           US   
9       The producer sources from two blocks of the vi...           US   
10      Elegance, complexity and structure come togeth...        Italy   
11      From 18-year-old vines, this supple well-balan...           US   
12      A standout even in this terrific lineup of 201...           US   
13      This wine is in peak condition. The tannins an...       France   
14      With its sophisticated mix of mineral, acid an...           US   
15      First made in 2006, this succulent luscious Ch...           US   
16      This blockbuster, powerhouse of a wine suggest...           US   
17      Nicely oaked blackberry, licorice, vanilla and...        Spain   
18      Coming from a seven-acre vineyard named after ...       France   
19      This fresh and lively medium-bodied wine is be...           US   
20      Heitz has made this stellar rosé from the rare...           US   
21      Alluring, complex and powerful aromas of grill...        Spain   
22      Tarry blackberry and cheesy oak aromas are app...        Spain   
23      The apogee of this ambitious winery's white wi...           US   
24      San Jose-based producer Adam Comartin heads 1,...           US   
25      Yields were down in 2015, but intensity is up,...  New Zealand   
26      Bergström has made a Shea designate since 2003...           US   
27      Focused and dense, this intense wine captures ...           US   
28      Cranberry, baked rhubarb, anise and crushed sl...           US   
29      This standout Rocks District wine brings earth...           US   
...                                                   ...          ...   
150889  A bizarre style of wine. The aromas are Port-l...           US   
150890  Wearing a heavy mantle of rather green, not-to...        Chile   
150892  A light, earthy wine, with violet, berry and t...           US   
150894  An attractive bouquet of pear, orange-lime and...        Chile   
150897  With a vegetal note, some cocoa and berry on t...        Chile   
150899  Basic Cabernet Sauvignon, with all its goods u...        Chile   
150900  Aromas of freshly cut lumber, complete with so...        Chile   
150901  Lavishly oaked, the fruit here struggles to ma...        Chile   
150902  This medium weight Chardonnay offered aromas o...        Chile   
150903  Very light berry and mint aromas open this aus...        Chile   
150905  There's not much point in making a reserve-sty...        Chile   
150906  This lovely wine, a Monopole, is already showi...       France   
150907  Rion holds back on the new oak, letting the pu...       France   
150908  Another premier cru from Michel Gros, this one...       France   
150909  This is a lovely, fragrant Burgundy, with a sm...       France   
150911  This needs a good bit of breathing time, then ...       France   
150912  The nose is dominated by the attractive scents...       France   
150914  Old-gold in color, and thick and syrupy. The a...           US   
150915  Decades ago, Beringer’s then-winemaker Myron N...           US   
150916  An impressive wine that presents a full bouque...           US   
150917  Light and elegant, this spicy, lively wine is ...       France   
150918  Jacquart makes a full-bodied, ripe style of Ch...       France   
150919  This classy example opens with a very flor

In [61]:
def transform_points_simplified(points):
    if points< 84:
        return 1
    elif points >= 84 and points < 88:
        return 2
    elif points >= 88 and points < 92:
        return 3
    elif points >= 92 and points < 96:
        return 4
    else:
        return 5
    
df1=df1.assign(points_simplified = df1['points'].apply(transform_points_simplified))
df1.head()

description country  \
0  This tremendous 100% varietal wine hails from ...      US   
1  Ripe aromas of fig, blackberry and cassis are ...   Spain   
2  Mac Watson honors the memory of a wine once ma...      US   
3  This spent 20 months in 30% new French oak, an...      US   
4  This is the top wine from La Bégude, named aft...  France   

                            designation             variety  price  \
0                     Martha's Vineyard  Cabernet Sauvignon  235.0   
1  Carodorum Selección Especial Reserva       Tinta de Toro  110.0   
2         Special Selected Late Harvest     Sauvignon Blanc   90.0   
3                               Reserve          Pinot Noir   65.0   
4                            La Brûlade  Provence red blend   66.0   

                    winery  points  points_simplified  
0                    Heitz      96                  5  
1  Bodega Carmen Rodríguez      96                  5  
2                 Macauley      96                  5  
3                    Ponzi      96                  5  
4     Domaine de la Bégude      95                  4

Description Vectorization:

In [26]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, classification_report


In [63]:
X=df1['description']
y=df1['points_simplified']
vectorizer=CountVectorizer()
vectorizer.fit(X)

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
        tokenizer=None, vocabulary=None)

In [64]:
X=vectorizer.transform(X)
print("Shape of sparse Matrix is", X.shape)
#print("Amount of non-zero occurrences is", X.nnz)


Shape of sparse Matrix is (94919, 26255)


Split the dataset into train and testing set

In [65]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=101)
rfc = RandomForestClassifier()
rfc.fit(X_train, y_train)

##testing model
predictions=rfc.predict(X_test)
print(classification_report(y_test,predictions))

/Users/aggrace/anaconda2/envs/py37/lib/python3.6/site-packages/sklearn/ensemble/forest.py:248: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


              precision    recall  f1-score   support

           1       0.91      0.70      0.79       715
           2       0.77      0.89      0.83      3520
           3       0.78      0.82      0.80      3734
           4       0.93      0.58      0.71      1447
           5       1.00      0.59      0.74        76

   micro avg       0.80      0.80      0.80      9492
   macro avg       0.88      0.71      0.77      9492
weighted avg       0.81      0.80      0.79      9492



Result with CountVectorizer: 81% precision, 80% recall and f1-score with 0.79 which are not good

Hence, I try different vectorizer(TF-IDF)

In [66]:
###Try TF-IDF
X=df1['description']
y=df1['points_simplified']

vectorizer=TfidfVectorizer()
vectorizer.fit(X)
X=vectorizer.transform(X)

In [67]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=101)
rfc = RandomForestClassifier()
rfc.fit(X_train, y_train)

# Testing model
predictions = rfc.predict(X_test)
print(classification_report(y_test, predictions))

/Users/aggrace/anaconda2/envs/py37/lib/python3.6/site-packages/sklearn/ensemble/forest.py:248: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


              precision    recall  f1-score   support

           1       0.92      0.66      0.77       715
           2       0.78      0.87      0.82      3520
           3       0.76      0.82      0.79      3734
           4       0.90      0.60      0.72      1447
           5       1.00      0.62      0.76        76

   micro avg       0.79      0.79      0.79      9492
   macro avg       0.87      0.71      0.77      9492
weighted avg       0.80      0.79      0.79      9492



Result with TF-IDF: 80% precision, 79% recall and f1-score with 0.79 which are not good as well

In [71]:
###Try logistic regression model as classifier
from sklearn.linear_model import LogisticRegression
LR_clf = LogisticRegression(random_state=101)
LR_clf.fit(X_train, y_train)
predictions=LR_clf.predict(X_test)
print(classification_report(y_test, predictions))

/Users/aggrace/anaconda2/envs/py37/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/aggrace/anaconda2/envs/py37/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:459: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


              precision    recall  f1-score   support

           1       0.84      0.46      0.60       715
           2       0.71      0.77      0.74      3520
           3       0.67      0.76      0.71      3734
           4       0.72      0.51      0.60      1447
           5       0.00      0.00      0.00        76

   micro avg       0.70      0.70      0.70      9492
   macro avg       0.59      0.50      0.53      9492
weighted avg       0.70      0.70      0.69      9492



/Users/aggrace/anaconda2/envs/py37/lib/python3.6/site-packages/sklearn/metrics/classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


precision of logistic regresion is 70%, and f1-score is 0.69, even worse than random forest